In [99]:
import pandas as pd
import numpy as np

## Lab - Data Frames 

# Name: Gabriella Rivera

**code2** 2-letter country code

**state** is missing for scandinavian countries

**type** the count type (Deaths or Confirmed)

**count** the count of the corresponding case types. This is a cumulative count, i.e. it is a count of all cases
    from 2020 onward and it never goes down.

**lockdown** the date when the country (largely) went to lockdown in 2020.

**population** in late 2020.

**countPC** count per capita

**growth** daily growth of the count

**growthPC** growth per capita

# 1 Loading Data and CSV Separator

In [100]:
df = pd.read_csv('covid-scandinavia.csv.bz2')
df

code2\tcountry\tstate\tdate\ttype\tcount\tlockdown\tpopulation\tcountPC\tgrowth\tgrowthPC
0     DK\tDenmark\t""\t2020-01-22\tConfirmed\t0\t202...                                       
1     DK\tDenmark\t""\t2020-01-22\tDeaths\t0\t2020-0...                                       
2     DK\tDenmark\t""\t2020-01-23\tConfirmed\t0\t202...                                       
3     DK\tDenmark\t""\t2020-01-23\tDeaths\t0\t2020-0...                                       
4     DK\tDenmark\t""\t2020-01-24\tConfirmed\t0\t202...                                       
...                                                 ...                                       
6523  SE\tSweden\t""\t2022-04-14\tDeaths\t18605\t\t1...                                       
6524  SE\tSweden\t""\t2022-04-15\tConfirmed\t2495996...                                       
6525  SE\tSweden\t""\t2022-04-15\tDeaths\t18605\t\t1...                                       
6526  SE\tSweden\t""\t2022-04-16\tConfirmed\t2495996...                                       
6527  SE\tSweden\t""\t2022-04-16\tDeaths\t18605\t\t1...                                       

[6528 rows x 1 columns]

In [101]:
# this isn't helpful because the data appears to be smashed into one column

In [102]:
df = pd.read_csv('covid-scandinavia.csv.bz2', sep="\t")
df

code2  country  state        date       type    count    lockdown  \
0       DK  Denmark    NaN  2020-01-22  Confirmed        0  2020-03-11   
1       DK  Denmark    NaN  2020-01-22     Deaths        0  2020-03-11   
2       DK  Denmark    NaN  2020-01-23  Confirmed        0  2020-03-11   
3       DK  Denmark    NaN  2020-01-23     Deaths        0  2020-03-11   
4       DK  Denmark    NaN  2020-01-24  Confirmed        0  2020-03-11   
...    ...      ...    ...         ...        ...      ...         ...   
6523    SE   Sweden    NaN  2022-04-14     Deaths    18605         NaN   
6524    SE   Sweden    NaN  2022-04-15  Confirmed  2495996         NaN   
6525    SE   Sweden    NaN  2022-04-15     Deaths    18605         NaN   
6526    SE   Sweden    NaN  2022-04-16  Confirmed  2495996         NaN   
6527    SE   Sweden    NaN  2022-04-16     Deaths    18605         NaN   

      population   countPC  growth  growthPC  
0        5837213  0.000000     NaN       NaN  
1        5837213  0.000000     NaN       NaN  
2        5837213  0.000000     0.0   0.00000  
3        5837213  0.000000     0.0   0.00000  
4        5837213  0.000000     0.0   0.00000  
...          ...       ...     ...       ...  
6523    10377781  0.001793    99.0   0.00001  
6524    10377781  0.240513     0.0   0.00000  
6525    10377781  0.001793     0.0   0.00000  
6526    10377781  0.240513     0.0   0.00000  
6527    10377781  0.001793     0.0   0.00000  

[6528 rows x 11 columns]

In [103]:
# now this data is separated into columns, as it should be

# 2 Data Summary

In [104]:
df.dtypes

code2          object
country        object
state         float64
date           object
type           object
count           int64
lockdown       object
population      int64
countPC       float64
growth        float64
growthPC      float64
dtype: object

In [105]:
subset = df[['country', 'date', 'type', 'count', 'population']]
subset
# df.drop(labels = ["code2", "state", "lockdown"], axis = 1)

country        date       type    count  population
0     Denmark  2020-01-22  Confirmed        0     5837213
1     Denmark  2020-01-22     Deaths        0     5837213
2     Denmark  2020-01-23  Confirmed        0     5837213
3     Denmark  2020-01-23     Deaths        0     5837213
4     Denmark  2020-01-24  Confirmed        0     5837213
...       ...         ...        ...      ...         ...
6523   Sweden  2022-04-14     Deaths    18605    10377781
6524   Sweden  2022-04-15  Confirmed  2495996    10377781
6525   Sweden  2022-04-15     Deaths    18605    10377781
6526   Sweden  2022-04-16  Confirmed  2495996    10377781
6527   Sweden  2022-04-16     Deaths    18605    10377781

[6528 rows x 5 columns]

# 3 What kind of data do we have here? Do these make sense?

(a) Find all the countries included in data. Are these all correct?

In [106]:
subset["country"].unique()

array(['Denmark', 'Finland', 'Norway', 'Sweden'], dtype=object)

This is a list of Scandinavian countries, all the sovereign states that is, except Iceland.  

(b) Find the minimum and maximum date. Are these realistic?

In [107]:
subset["date"].min()

'2020-01-22'

In [108]:
subset["date"].max()

'2022-04-16'

Yes, these maximum and minimum dateas are realistic.

(c) Find the minimum and maximum count. Are these realistic?

In [109]:
subset["count"].min()

0

In [110]:
subset["count"].max()
# max_count_id = subset["count"].idxmax()
# subset.loc[max_count_id, "country"]

3097015

The minimum count is realistic, but the maximum count isn't.

In [111]:
subset.count.min()

AttributeError: 'function' object has no attribute 'min'

I can compute it as   ["count"].min()   but I cannot compute this as   .count.min()   because subset.count is attempting to access a method or attribute named count on the DataFrame object. In pandas, DataFrame.count is a method that returns the count of non-NA/null observations over the requested axis. So, if we are to call subset.count(), it would return a Series with the number of non-NA/null values in each column of the DataFrame.

However, if we do not follow count with parentheses, like this: subset.count, it won't execute the count() method, but will instead return a reference to the method itself, not the count of values. We cannot call min() on a reference to a method. 

# 3 Text Question

1. Why did you choose to take info370? What do you think, what are the most valuable skills you
hope to learn here?

I chose to take INFO 370 because I took CSE 163 and enjoyed it. I am looking to enter a career in data analytics and this class is a step in that direction. I hope to dive deeper into the statistical side of analytics. 

# 4 Cases per capita (not graded)

1. Filter the data frame you created above and keep only the confirmed cases.

In [ ]:
confirmed = subset[subset['type'] == 'Confirmed']
confirmed

country        date       type    count  population
0     Denmark  2020-01-22  Confirmed        0     5837213
2     Denmark  2020-01-23  Confirmed        0     5837213
4     Denmark  2020-01-24  Confirmed        0     5837213
6     Denmark  2020-01-25  Confirmed        0     5837213
8     Denmark  2020-01-26  Confirmed        0     5837213
...       ...         ...        ...      ...         ...
6518   Sweden  2022-04-12  Confirmed  2491980    10377781
6520   Sweden  2022-04-13  Confirmed  2491980    10377781
6522   Sweden  2022-04-14  Confirmed  2495996    10377781
6524   Sweden  2022-04-15  Confirmed  2495996    10377781
6526   Sweden  2022-04-16  Confirmed  2495996    10377781

[3264 rows x 5 columns]

How many cases do you have in the subset?
3264

2. Create a new variable, confirmed cases per capita, by dividing the confirmed cases’ count by population.

In [ ]:
confirmed["confirmed_cases_per_capita"] = confirmed["count"]/confirmed["population"]

C:\Users\Gabriella Rivera\AppData\Local\Temp\ipykernel_13536\193692614.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  confirmed["confirmed_cases_per_capita"] = confirmed["count"]/confirmed["population"]


3. You may notice a warning: “A value is trying to be set on a copy of a slice...”. Why do we get this
warning? How can we get rid of it? Repeat the previous steps in a way that you do not get the
warning!

We need to make an explicit copy with .copy method before we start any modifications:

In [ ]:
confirmed = subset[subset['type'] == 'Confirmed'].copy()     # explicit copy
confirmed["confirmed_cases_per_capita"] = confirmed["count"]/confirmed["population"]
confirmed

country        date       type    count  population  \
0     Denmark  2020-01-22  Confirmed        0     5837213   
2     Denmark  2020-01-23  Confirmed        0     5837213   
4     Denmark  2020-01-24  Confirmed        0     5837213   
6     Denmark  2020-01-25  Confirmed        0     5837213   
8     Denmark  2020-01-26  Confirmed        0     5837213   
...       ...         ...        ...      ...         ...   
6518   Sweden  2022-04-12  Confirmed  2491980    10377781   
6520   Sweden  2022-04-13  Confirmed  2491980    10377781   
6522   Sweden  2022-04-14  Confirmed  2495996    10377781   
6524   Sweden  2022-04-15  Confirmed  2495996    10377781   
6526   Sweden  2022-04-16  Confirmed  2495996    10377781   

      confirmed_cases_per_capita  
0                       0.000000  
2                       0.000000  
4                       0.000000  
6                       0.000000  
8                       0.000000  
...                          ...  
6518                    0.240126  
6520                    0.240126  
6522                    0.240513  
6524                    0.240513  
6526                    0.240513  

[3264 rows x 6 columns]

4. Now extract the cases per capita variable your created for Sweden and Norway. You should create
two new workspace variables and assign the corresponding columns to those.

In [ ]:
# confirmed["country"] == "Sweden"
# Sweden = confirmed.loc[confirmed["country"] == "Sweden", ["country", "confirmed_cases_per_capita"]]
Sweden = confirmed[confirmed["country"] == "Sweden"]["confirmed_cases_per_capita"]
# Sweden = confirmed.loc[confirmed["country"] == "Sweden", ["confirmed_cases_per_capita"]]
Sweden

4896    0.000000
4898    0.000000
4900    0.000000
4902    0.000000
4904    0.000000
          ...   
6518    0.240126
6520    0.240126
6522    0.240513
6524    0.240513
6526    0.240513
Name: confirmed_cases_per_capita, Length: 816, dtype: float64

In [ ]:
# Norway = confirmed[confirmed["country"] == "Norway"]["confirmed_cases_per_capita"]
Norway = confirmed.loc[confirmed["country"] == "Norway", ["confirmed_cases_per_capita"]]
Norway

confirmed_cases_per_capita
3264                    0.000000
3266                    0.000000
3268                    0.000000
3270                    0.000000
3272                    0.000000
...                          ...
4886                    0.263276
4888                    0.263425
4890                    0.263496
4892                    0.263548
4894                    0.263625

[816 rows x 1 columns]

In [ ]:
print(type(Sweden))
print(type(Norway))

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


The type of the variable depends on how you create the variable